core/load_copyright_map.py

```diff
@@ -1,34 +1,60 @@
 import logging
 import sys
 import datetime
+import time
+import pdb
 from core.models import Copyright
 from core.models import LccnDateCopyright
+from django.db import transaction
+from django.core.validators import URLValidator
+from django.core.exceptions import ValidationError
 
-def check_uri(_uri):
-    try:
-        result = Copyright.objects.filter(uri=_uri)
-        return result.exists()
-    except Exception as e:
-        _logger.exception(e)
 
 def loadCopyrightMap(inputfilename):
 
     _logger = logging.getLogger(__name__)
+    val = URLValidator()
+
     try:
         f = open(inputfilename, 'r')
-        for line in f:
-            arr = line.split("\t")
-            if len(arr) == 4:
-                if check_uri(arr[3].strip()):
-                    record = LccnDateCopyright()
-                    record.lccn = arr[0]
-                    record.start_date = (arr[1])
-                    record.end_date = (arr[2])
-                    result = Copyright.objects.filter(uri=arr[3].strip())
-                    record.uri = result[0]
-                    record.save()
+    except IOError as e:
+        _logger.exception(e)
+        sys.exit(1)
+
+    for line in f:
+        arr = line.split("\t")
+        if len(arr) == 4:
+            try:
+                val(arr[3].strip())
+            except ValidationError as e:
+                _logger.exception(e)
+                sys.exit(1)
+            try:
+                sdate = datetime.datetime.strptime(arr[1], "%Y-%m-%d").date()
+                edate = datetime.datetime.strptime(arr[2], "%Y-%m-%d").date()
+            except ValueError as e:
+                _logger.exception(e)
+                sys.exit(1)
 
-    except Exception as e:
-       _logger.exception(e)
+            record = LccnDateCopyright()
+            record.lccn = arr[0]
+            record.start_date = sdate
+            record.end_date = edate
 
+            try:
+                result = Copyright.objects.filter(uri=arr[3].strip())
+                record.copyright = result[0]
+            except Exception as e:
+                _logger.exception(e)
+                sys.exit(1)
+
+            try:
+                with transaction.atomic():
```

Comment:
Same as above - wrap the whole `for` loop with the transaction to make sure the whole group loads successfully or else nothing is loaded.